### ISSUE TITLE
[Feature Request] Enable multiple certURLs

### ISSUE THREAD

**Issue by PrplHaz4** [Ylianst/MeshCentral#3798]
Enable multiple certURLs to allow automatic configuration for split-horizon dns or multiple tls proxies in front of meshcentral.

This would resolve numerous outstanding issues related to CloudFlare connectivity.  Most recently (?), https://github.com/Ylianst/MeshCentral/issues/3580

Simple use case:
- Clients outside network go through CloudFlare Proxy
- Clients inside network go through internal reverse Proxy

This would allow internet-facing clients to connect and be aware of only the CF proxy, while internal clients can connect directly to your internal reverse proxy.

There may be a more elegant solution to this problem, but it seems common enough, and IMO allowing meshcentral to hide behind CF is a great way to mitigate the risk of exposing a service like this to the public internet.

Keep up the great work - you guys really deliver a truly remarkable OSS product!

**Comment by Ylianst** [comment]
I do agree that having MeshCentral behind Cloud Flare is a good security measure. If I understand correct, your asking that MeshCentral be allowed to be behind multiple TLS certificates. So, you could have MeshCentral behind a reverse proxy using a LE cert? and a different reverse proxy using a Amazon AWS cert?

The purpose of "CertURL" is to tell MeshCentral what TLS certificate it's going to be behind. Having multiple "CerlURL" would allows for multiple certs. Is that correct?

**Comment by PrplHaz4** [comment]
> I do agree that having MeshCentral behind Cloud Flare is a good security measure. If I understand correct, your asking that MeshCentral be allowed to be behind multiple TLS certificates. So, you could have MeshCentral behind a reverse proxy using a LE cert? and a different reverse proxy using a Amazon AWS cert?
> 
> The purpose of "CertURL" is to tell MeshCentral what TLS certificate it's going to be behind. Having multiple "CerlURL" would allows for multiple certs. Is that correct?

Correct on all fronts (from my point of view at least)!

**Comment by PrplHaz4** [comment]
> I do agree that having MeshCentral behind Cloud Flare is a good security measure. If I understand correct, your asking that MeshCentral be allowed to be behind multiple TLS certificates. So, you could have MeshCentral behind a reverse proxy using a LE cert? and a different reverse proxy using a Amazon AWS cert?
> 
> The purpose of "CertURL" is to tell MeshCentral what TLS certificate it's going to be behind. Having multiple "CerlURL" would allows for multiple certs. Is that correct?

Quick diagram of this proposal:
![image](https://user-images.githubusercontent.com/841414/160144318-5df78063-b993-4ebf-8504-9a25d184777a.png)

**Comment by aj-bayanat** [comment]
Hi, I have the same problem that this feature would solve. I am running MC behind Caddy with a cloudflare cert and a internal DNS that points to the Caddy server. 
I do NOT want to port forward on my work network, so instead I opted for using cloudflare tunnels. (Now thats its free for everyone, yay! ). I get the same mismatch of SHA certs for clients trying to connect from outside my network.
One way I tried to solve this was to remove my internal dns and only use cloudflare however this results in the error
Failed to load certs from mesh.mydomain.com. 
I suspect this is because my internal dns (windows server) has to route my domain by using a cname. This is because I have many other internal web applications so I need to route mydomain.com through my internal DNS. 
Usually to bypass this for some webapps the internal DNS is set to point to CNAME, so for example mesh.mydomain.com.cdn.cloudflare.net, this usually works but for some reason MC is not able to grab the cert when I do this. 
Any advice would be greatly appreciated.

**Comment by twistedsanity** [comment]
I'm also having this issue split cloud flare and local. Any progress would be amazing.

**Comment by si458** [comment]
@twistedsanity can u explain a little more for me plz about ur setup etc?

I'm trying to figure out why u would have split dns?
Surely ur meshcentral dns is just ur reverse proxy ip on the Internet? And internal devices do a hairpin nat (miktorik terms) so it can connect to the reverse proxy internally even tho u use its external ip?

**Comment by Renaud11232** [comment]
Hello,

Adding my 2 cents here : 

When using Cloudflare, your DNS records, even tho they can, most likely don't point to your actual IP but rather to CF's reverse proxies, which then forwards requests to your actual server IP.

Without split horizon DNS, accessing your hosted services (such as MeshCentral) from your local network would cause un-needed traffic from your public IP to Cloudflare's servers and back, just to access resources that actually are on the same local network.

**Comment by si458** [comment]
@PrplHaz4 so does ur internal dns resolve the IP of ur meshcentral to the the reverse proxy internally, and then everyone outside would get the meshcentral ip of cloudflare which in turn tunnels it to reverse proxy?

Just trying to get an idea, so I can setup environment and get testing!

**Comment by twistedsanity** [comment]
@Renaud11232 explained the issue perfectly. 
Also the diagram @PrplHaz4 provided is almost identical to my setup.

I run meshcentral behind an reverse proxy (HAProxy) which services clients locally. This same HAProxy server also sits behind cloudflare, which I use for security, protection and a few other features. I really do not want to forward internal clients out via the internet 'hairpinning', therefore split DNS allows me to keep all internal traffic internal. Also due to the fantastic and free protections cloudflare offers is its not something I would wish to disable.

The actual issue is that the two different routes use two different certificates, even though they are both perfectly valid. My internal clients get presented and try to use a lets encrypt cert automatically managed by acme certificates (let's encrypt) for my external url (wildcard). The external route via cloudflare uses its own automatically managed (by CF) let's encrypt cert which is also for my same external url (wildcard). With cloudflare unless you pay silly money for a large enterprise grade account and have the knowledge to manage all cert changes programmatically this is an unchangeable scenario unfortunately.

@si458 I would be more than happy to jump on a session and share some config I wouldn't share publicly if it helps at all.

**Comment by si458** [comment]
@twistedsanity right thank u for that! It does explain it and I understand now!
So in theory as the issue says u need certurl to be able to have multiple set so it can then verify with 1 cert or another! 

I will have a look at source code over the weekend!

**Comment by twistedsanity** [comment]
Just wondered if anyone had a look at the possibility/feasibility of being able to add multiple certs? If not I might have to temporarily move away until it becomes an option, as running multiple instances is not really an option for me.

**Comment by PrplHaz4** [comment]
> @PrplHaz4 so does ur internal dns resolve the IP of ur meshcentral to the the reverse proxy internally, and then everyone outside would get the meshcentral ip of cloudflare which in turn tunnels it to reverse proxy?
> 
> Just trying to get an idea, so I can setup environment and get testing!

Sorry for not responding!  Yes, that is how it works - internal network clients use internal DNS to resolve and hit the RP IP directly, external clients use unknown DNS server to resolve the URL to cloudflare IPs (which then point to RP IP).

**Comment by bhuvicp** [comment]
Thanks @si458 for taking a look at this! Wondering if you might be able to provide any insights as to how easy/hard it would be to implement this feature.

**Comment by raregtp** [comment]
Just chiming in that I have a similar split DNS situation that, while I can script a solution to ensure I'm using the same certificate at both entry points, it would be MUCH easier if I could just have multiple certURL entries or a way to trust the LetsEncrypt certificate that my edge proxy automation has set up.

**Comment by ckocyigit** [comment]
I have the same situation as @raregtp

**Comment by twistedsanity** [comment]
Hopefully this will see some movement in the new year as this is still quite a major issue for me and I hope I can resolve before I plan to swap out the current firewall as I am not sure how to do it with the new one.

**Comment by si458** [comment]
@twistedsanity I will indeed look into it next year! It should be simple enough, but when I first looked at it, it was all over the place as the checkers for signature etc only checked a value, where as what we need to do is to check the array and if a signature is there then move on

**Comment by made-indrayana** [comment]
Happy new year! First I want to thank you all for your hard works :) I have the exact same situation as @PrplHaz4. Just leaving a comment here to know if this is done 👍

**Comment by si458** [comment]
@made-indrayana still in the works!
got a lot of feature requests and bug fixes on my plate to sort this year!
but its on the to-do list!

**Comment by PrplHaz4** [comment]
not stale!

**Comment by twistedsanity** [comment]
100% not stale, just waiting for an update.

**Comment by twistedsanity** [comment]
Not stale just patiently waiting for an update.

**Comment by Deenyoro** [comment]
Not stale still need this.

**Comment by twistedsanity** [comment]
Not stale still need this.

**Comment by si458** [comment]
yes ive been looking into this in the background and its not as straight forward as i hoped

we can change the certurl to an array no problem and get meshcentral to load those certs no problem
the problem is with the meshagent, because it hashes the certificates and includes those in the exe and the .msh
and the meshagent only supports 1 hash not multiple
so it needs a meshagent rewrite 👎 

edit: its still on the Todo list as you see from projects on github haha

**Comment by PrplHaz4** [comment]
> yes ive been looking into this in the background and its not as straight forward as i hoped
> 
> we can change the certurl to an array no problem and get meshcentral to load those certs no problem the problem is with the meshagent, because it hashes the certificates and includes those in the exe and the .msh and the meshagent only supports 1 hash not multiple so it needs a meshagent rewrite 👎
> 
> edit: its still on the Todo list as you see from projects on github haha

Doesn't it only hash the first certificate it sees when it tries to connect?  I think this behavior would still work as most machines are either routed through your lan or not - rarely will they be both (or change)...

Edit to add: IOW, meshagent (I think?) would really only need to support one certificate at a time, as long as it's the "right" one

**Comment by durzo** [comment]
how does the current mechanism work with LetsEncrypt? wont the certificate hash change on renewal? which is currently 90 days and going to 45 days in the near future.

Does this mean we need to restart the mesh central server to fetch the renewed certUrl on each certificate renewal?
What about updating agent keys on each certificate renewal?

Would a possible solution here be, instead of checking the cert hash, have the server generate some unique signature which agents would check for instead, and have the extent of ssl cert verification simply check if its a valid cert and not expired?

You could just use PGP for this, have the server embed its public key into the agent and spit out a signed message via https on agent check-in to verify the server is still trusted.

**Comment by durzo** [comment]
FWIW I have found a workaround that works for me, allowing me to use 2 sets of DNS that differ among internal vs external hosts.

Key configuration:
```
  "settings": {
    "cert": "meshcentral.myexternaldomain.com",
    "tlsOffload": false,
    "port": 443,
    "trustedproxy": "192.168.1.13,CloudFlare",
    "WANonly": true,
    "ignoreAgentHashCheck": "192.168.1.0/24"
  },
  "domains": {
    "": {
      "certUrl": "https://meshsslcheck.myexternaldomain.com",
    }
  }
```

The setup:
* meshcentral.myexternaldomain.com resolves internally (LAN) as 192.168.1.13, which is my reverse proxy (NPM) using a LetsEncrypt SSL certificate generated by DNS challenge, not HTTP. NPM connects to MeshCentral via https (MC's self signed cert).
* meshcentral.myexternaldomain.com resolves externally (internet) as my cloudflare tunnel, but this could be your WAN IP just as well.
* meshsslcheck.myexternaldomain.com resolves both internally and externally as the cloudflare tunnel / WAN IP

When the meshcentral server starts up, it gets the external ssl certificate provided by Cloudflare from the meshsslcheck hostname, which is what all external agents see for meshcentral.mydomain.com.

Internal agents do not complain about the certificate mismatch due to `ignoreAgentHashCheck` being set to the LAN subnet. If those devices leave the internal network, they correctly verify the external certificate over the internet.

This allows me to have end to end SSL communication for both internal and external agents

It's not perfect, but it works for me.

**Comment by Suncatcher** [comment]
I want to access the Meshcentral server from behind the corporate firewall, my server is IPv6-only, and my corporate firewall doesn't like IPv6 DNS resolution, I cannot reach it. On the server I decided to setup a passthru Cloudflare tunnel like that:

    cloudflared tunnel --url http://localhost:80

However, trying to reach the tunnel URL from the firewall machine I get 

>cloudflareurl.com redirected you too many times

So, I assume, what's going on:

1. Cloudflare makes a tunnel on HTTP port
2. It translated IPv4 requests to IPv6 and trying to make it HTTP
3. Mesh receives the HTTP request and trying to forward it to HTTPS
4. Cloudflare received HTTPS back and trying to redirect it to port 80 again
5. Infinite loop

My assumption is confirmed by curl response:

    user#/ curl cloudflareurl.com
    user#/ curl -> found. Redirecting to https://cloudflareurl.com:443/ 

Will this FR, that you are discussing here, solve my problem, or it can be solved by the the existing Meshcentral settings or features? As I understand this feature request is to be able for Meshcentral to support several certificate domains. Is it?
My requirement is kinda the same, I want Mesh to automatically switch to HTTPS all requests coming to 80 port, and I want to disable that for reverse proxy requests like from Cloudflare tunnel.

**Comment by si458** [comment]
@Suncatcher no, you need to point the cloudflare tunnel to port 443 instead of port 80
Also if using cloudflare u need to add in `trustedproxy: "cloudflare"` to get the correct ip address of remote devices AND use https instead of http

**Comment by Suncatcher** [comment]
> Also if using cloudflare u need to add in `trustedproxy: "cloudflare"` to get the correct ip address of remote devices AND use https instead of http

I made the corrections like you said, now I get the following error:

    ERR  error="Unable to reach the origin service. The service may be down or it may not be responding to traffic from cloudflared: tls: failed to verify certificate: x509: certificate is valid for mymeshcentralserver.com, not localhost" connIndex=0 event=1 ingressRule=0 originService=https://localhost:443

I'm starting tunnel like 

    cloudflared tunnel --url https://localhost:443

should I start another way?

**Comment by si458** [comment]
@Suncatcher the error in in the output ` failed to verify certificate: x509: certificate is valid for mymeshcentralserver.com, not localhost`
you need to set cloudflare to ignore the ssl cert and not verify it
or if you want you can add `tlsoffload: '127.0.0.1` and `trustedProxy: 'cloudflare'` to config.json then restart meshcentral
then change the cloudflared to use http://localhost:443 instead (only do this really if cloudflared is running on the same machine thats running meshcentral as it would transfer stuff in plain text between the cloudflare binary and meshcentral)

**Comment by Suncatcher** [comment]
> you need to set cloudflare to ignore the ssl cert and not verify it

okay, great. that worked. running the tunnel with `--no-tls-verify` did the trick. I didn't try the second option with `tlsoffload: '127.0.0.1`. What is more secure and more architecturally correct, `tlsoffload` or `--no-tls-verify`? Or they are basically the same?

**Comment by si458** [comment]
@Suncatcher both the same really just depends what your org wants security wise

<img width="1020" height="570" alt="Image" src="https://github.com/user-attachments/assets/1d8c67c8-c1d6-43ef-8c5f-bd30f3e5e6ed" />

**Comment by Suncatcher** [comment]
@si458 could you please check the issue here https://github.com/Ylianst/MeshCentral/discussions/7884?

**Comment by MobileManiC** [comment]
Currently seeing this issue when using a Reverse proxy that has multiple HA nodes, but each proxy node is using individually generated certificates. This seems to be currently unresolvable considering how Agents verify the server (only allowing a single possible cert per server).

"certUrl" function will load just one of these (from the node that was just active at that particular moment), cert from other node will not be trusted => with two proxy nodes, that is 50% chance of successful connect each time 🤷 

😢  Any ideas here? Looks like rewriting Agent verification to stop being based on cert hashes seems like the best solution, but complex...

**Comment by si458** [comment]
@MobileManiC the problem we have at the minute is that meshcentral encodes some of the hashs from the cert into the meshagent exe
and sadly the meshagent ONLY supports reading 1 cert, so it would require an agent update to support reading 2 certificates
i havent had time to check into it properly but i will try again when i have chance!

**Comment by xanpri** [comment]
Hello,

I am also experiencing the same issue. I actually have an existing issue related to this topic: **https://github.com/Ylianst/MeshAgent/issues/371**.

In our corporate environment, we need to configure two addresses in the MeshAgent because we have two reverse proxies: one external proxy, used for Internet connections and accessible through a public address, and another internal proxy, used for connections within our corporate network.

Our DNS policy does not allow us to use the public address as an alias on our internal DNS server.

Regarding the certificate, I believe we can work around this issue in two ways: by using the same certificate on both reverse proxies (internal and external), or by configuring TCP Pass-Through on the reverse proxies, allowing MeshCentral itself to handle certificate validation.

With the second option, we could configure the agent DNS aliases as follows:

`"agentAliasDNS": [ "myserver.mydomain.com", "agents.myserver.mydomain.local" ]`

and use:

`"agentPort": 4433`

while keeping the certificate configured as:

`"cert": "myserver.mydomain.com"`

This would allow us to configure TCP Pass-Through on the reverse proxies specifically for agent connections, similarly to how connections via AMT/CIRA work.

Would it be possible to provide a MeshAgent build that we could use for initial testing with both addresses, even if the issue regarding the use of two certificates in the MeshAgent has not yet been fully resolved?

I believe this would allow us to validate this approach in our corporate environment and determine whether it meets our connectivity and security requirements.

If appropriate, this comment could also be considered as related to **https://github.com/Ylianst/MeshAgent/issues/371**, since the two issues are closely connected.

Thank you!
